pandas/io/gcs.py

```diff
@@ -0,0 +1,26 @@
+""" GCS support for remote file interactivity """
+from pandas import compat
+try:
+    from google.auth.exceptions import GoogleAuthError
+    import gcsfs
+except ImportError:
+    raise ImportError("The gcsfs library is required to handle GCS files")
+
+
+def get_filepath_or_buffer(filepath_or_buffer, encoding=None,
+                           compression=None, mode=None):
+
+    if mode is None:
+        mode = 'rb'
+
+    gcsfs_logger_disabled = gcsfs.core.logger.disabled
+    gcsfs.core.logger.disabled = True
+    fs = gcsfs.GCSFileSystem()
+    try:
+        filepath_or_buffer = fs.open(filepath_or_buffer, mode)
+    except (compat.FileNotFoundError, GoogleAuthError, gcsfs.utils.HtmlError):
+        fs = gcsfs.GCSFileSystem(token='anon')
```

Comment:
note: this was originally a hack for not breaking backwards compat with pre-s3fs pandas. I'd like to eventually move to something like dask's `store_options` here, a dict of keywords that gets passed down to the backend (https://github.com/pandas-dev/pandas/issues/19904)

Reply:
I agree this would be the best way to do things. For now I simplified things and am just using `gcsfs`'s default authentication preferences. If allowing anonymous access to public files is important we could try to think of a better way to accomplish that, but arguably `gcsfs` should handle that logic?

Reply:
Note that gcsfs defaults to anon access if no other mechanisms worked - but this will not happen if one of the other mechanisms contained real credentials that turned out to be no longer valid.

Reply:
That's what I thought but it didn't seem to be the case when I tested it. In a `python:3.6.4` Docker image w/ `gcsfs` installed I get:
```
# default auth fails
In [2]: gcsfs.GCSFileSystem().open('gs://gcp-public-data-landsat/index.csv.gz').read(10)
# lots of error logs from failing to access /computeMetadata/v1/instance/service-accounts/default/?recursive=true
...
RefreshError: HTTPConnectionPool(host='metadata.google.internal', port=80): Max retries exceeded with url: /computeMetadata/v1/instance/service-accounts/default/?recursive=true (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fab36881518>: Failed to establish a new connection: [Errno -2] Name or service not known',))

# specifying token='anon' works
In [3]: gcsfs.GCSFileSystem(token='anon').open('gs://gcp-public-data-landsat/index.csv.gz').read(10)
Out[3]: b'\x1f\x8b\x08\x08\x89\xd8\xd6Z\x02\xff'
```

Should I create a separate `gcsfs` issue for this?

Reply:
Hm, OK, it should be an issue - but worth checking whether it's still the case with current master.